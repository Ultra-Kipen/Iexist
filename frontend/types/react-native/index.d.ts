// types/react-native/index.d.ts

// 이미 존재하는 모듈을 확장하는 방식으로 변경
import 'react-native';

// 타입스크립트에게 이 파일이 타입 모듈임을 알려줌
export {};

// 타입 확장을 위한 전역 선언
declare global {
  // React Native에 관련된 글로벌 타입을 필요에 따라 추가
}

// 모듈 보강(augmentation) 방식으로 변경
// 이미 정의된 인터페이스나 타입에 새로운 정의를 추가
declare module 'react-native/Libraries/Animated/Animated' {
  // 필요한 타입 정의가 있다면 여기에 추가
}

// 실제로는 아무 것도 선언하지 않고, 기존 타입 시스템을 유지하는 방식
// 이렇게 하면 TypeScript 오류가 사라질 수 있음