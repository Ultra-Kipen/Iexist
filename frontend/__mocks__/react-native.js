// __mocks__/react-native.js
'use strict';

// 실제 react-native를 가져오려고 시도하지 않고, 모든 것을 모킹합니다
const DevMenuMock = {
  show: jest.fn(),
  debugRemotely: jest.fn(),
};

module.exports = {
  // 기본 컴포넌트
  View: 'View',
  Text: 'Text',
  TouchableOpacity: 'TouchableOpacity',
  Image: 'Image',
  ScrollView: 'ScrollView',
  TextInput: 'TextInput',
  
  // TurboModuleRegistry 모킹
  TurboModuleRegistry: {
    get: jest.fn(() => null),
    getEnforcing: jest.fn((name) => {
      if (name === 'DevMenu') {
        return DevMenuMock;
      }
      return null;
    })
  },
  
  // DevMenu 모킹
  DevMenu: DevMenuMock,
  
  // StyleSheet 모킹
  StyleSheet: {
    create: jest.fn((styles) => styles),
    flatten: jest.fn((style) => style),
    hairlineWidth: 1,
  },
  
  // Platform 모킹
  Platform: {
    OS: 'android',
    select: jest.fn((obj) => obj.android || obj.default),
    Version: {
      SDK_INT: 28
    }
  },
  
  // Dimensions 모킹
  Dimensions: {
    get: jest.fn(() => ({ width: 375, height: 667 })),
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
  },
  
  // Alert 모킹
  Alert: {
    alert: jest.fn(),
  },
  
  // 네이티브 모듈 모킹
  NativeModules: {
    DevMenu: DevMenuMock,
    NativeAnimatedHelper: {
      startAnimatingNode: jest.fn(),
      stopAnimation: jest.fn(),
    },
    RNCNetInfo: {
      getCurrentState: jest.fn(() => Promise.resolve()),
      addListener: jest.fn(),
      removeListeners: jest.fn(),
    },
    StatusBarManager: {
      HEIGHT: 42,
      setStyle: jest.fn(),
      setHidden: jest.fn(),
    }
  },
  
  // Animated 모킹
  Animated: {
    Value: jest.fn(() => ({
      setValue: jest.fn(),
      interpolate: jest.fn(() => ({
        interpolate: jest.fn(),
      })),
    })),
    timing: jest.fn(() => ({
      start: jest.fn(callback => callback && callback({ finished: true })),
    })),
    spring: jest.fn(() => ({
      start: jest.fn(callback => callback && callback({ finished: true })),
    })),
    View: 'Animated.View',
    Text: 'Animated.Text',
    Image: 'Animated.Image',
    createAnimatedComponent: jest.fn(comp => comp),
  },
  
  // 그 외 자주 사용되는 API
  Keyboard: {
    dismiss: jest.fn(),
    addListener: jest.fn(() => ({ remove: jest.fn() })),
  },
  
  LayoutAnimation: {
    configureNext: jest.fn(),
    create: jest.fn(),
    Types: {},
    Properties: {},
  },
  
  UIManager: {
    measure: jest.fn(),
    measureInWindow: jest.fn(),
  },
  
  NativeEventEmitter: jest.fn(() => ({
    addListener: jest.fn(),
    removeListeners: jest.fn(),
  })),
  
  useWindowDimensions: jest.fn(() => ({ width: 375, height: 667 })),
  
  PixelRatio: {
    get: jest.fn(() => 2),
    getFontScale: jest.fn(() => 1),
    getPixelSizeForLayoutSize: jest.fn(size => size * 2),
    roundToNearestPixel: jest.fn(size => size),
  },
  
  BackHandler: {
    addEventListener: jest.fn(() => ({ remove: jest.fn() })),
    removeEventListener: jest.fn(),
  },
  
  AppState: {
    addEventListener: jest.fn(() => ({ remove: jest.fn() })),
    removeEventListener: jest.fn(),
    currentState: 'active',
  },
  
  // 더 많은 API 모킹이 필요하면 여기에 추가할 수 있습니다
};