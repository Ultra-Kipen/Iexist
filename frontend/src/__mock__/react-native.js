// src/__mocks__/react-native.js
export const Alert = {
    alert: jest.fn()
  };
  
  export const View = jest.fn().mockImplementation(({children, ...props}) => children);
  export const Text = jest.fn().mockImplementation(({children, ...props}) => children);
  export const TouchableOpacity = jest.fn().mockImplementation(({children, ...props}) => children);
  export const Image = jest.fn().mockImplementation(({source, ...props}) => null);
  export const ScrollView = jest.fn().mockImplementation(({children, ...props}) => children);
  export const TextInput = jest.fn().mockImplementation(({...props}) => null);
  
  export default {
    Alert,
    View,
    Text,
    TouchableOpacity,
    Image,
    ScrollView,
    TextInput
  };