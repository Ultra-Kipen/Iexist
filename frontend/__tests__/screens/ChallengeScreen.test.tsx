import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { ScrollView, Text, View, TouchableOpacity } from 'react-native';

// Jest mock은 모듈 외부 변수를 참조할 수 없으므로 내부에서 React 재사용
jest.mock('react-native-paper', () => {
  const React = require('react');
  const { View, Text, TouchableOpacity } = require('react-native');
  
  return {
    useTheme: () => ({
      colors: {
        primary: '#6200ee',
      },
    }),
    Card: {
      Content: (props: any) => React.createElement(View, {}, props.children),
      Actions: (props: any) => React.createElement(View, {}, props.children),
    },
    Title: (props: any) => React.createElement(Text, {}, props.children),
    Paragraph: (props: any) => React.createElement(Text, {}, props.children),
    Button: (props: any) => React.createElement(TouchableOpacity, { onPress: props.onPress }, React.createElement(Text, {}, props.children)),
    ProgressBar: (props: any) => React.createElement(View, { 
      accessibilityRole: 'progressbar', 
      progress: props.progress
    }),
    List: {
      Item: (props: any) => React.createElement(View, {}, React.createElement(Text, {}, props.title)),
      Icon: () => React.createElement(View, {}, null),
    },
  };
});

// ChallengeScreen 컴포넌트 모킹
const ChallengeScreen = () => {
  const challenges = [
    {
      id: 1,
      title: '7일간의 감사 일기',
      description: '매일 감사한 일 3가지를 기록해보세요.',
      participants: 128,
      duration: 7,
      progress: 0.4,
    },
    {
      id: 2,
      title: '30일 긍정 에너지 나누기',
      description: '하루에 한 번 주변 사람에게 긍정적인 말을 해보세요.',
      participants: 56,
      duration: 30,
      progress: 0.2,
    },
    {
      id: 3,
      title: '21일 명상 습관 만들기',
      description: '매일 10분씩 명상을 하고 느낀 점을 공유해보세요.',
      participants: 89,
      duration: 21,
      progress: 0.6,
    },
  ];

  const handleJoinChallenge = (challengeId: number) => {
    console.log('Joining challenge:', challengeId);
  };

  // React Native Paper 컴포넌트 사용
  const { Card, Title, Paragraph, Button, ProgressBar, List } = require('react-native-paper');
  const theme = { colors: { primary: '#6200ee' } };

  return (
    <ScrollView>
      <Title>현재 진행 중인 챌린지</Title>
      {challenges.map((challenge) => (
        <View key={challenge.id} style={{ marginBottom: 16 }}>
          <View>
            <Title>{challenge.title}</Title>
            <Paragraph>{challenge.description}</Paragraph>
            <View>
              <ProgressBar progress={challenge.progress} color={theme.colors.primary} />
              <Paragraph>{`${Math.round(challenge.progress * 100)}% 완료`}</Paragraph>
            </View>
            <List.Item
              title={`참여자: ${challenge.participants}명`}
              left={(props: any) => <List.Icon {...props} icon="account-group" />}
            />
            <List.Item
              title={`기간: ${challenge.duration}일`}
              left={(props: any) => <List.Icon {...props} icon="calendar-range" />}
            />
          </View>
          <View>
            <Button onPress={() => handleJoinChallenge(challenge.id)}>참여하기</Button>
          </View>
        </View>
      ))}
    </ScrollView>
  );
};

// Mock console.log to track calls
const mockConsoleLog = jest.fn();
console.log = mockConsoleLog;

describe('ChallengeScreen', () => {
  beforeEach(() => {
    mockConsoleLog.mockClear();
  });

  it('renders correctly with challenges', () => {
    const { getByText, getAllByText } = render(<ChallengeScreen />);

    // Check title is rendered
    expect(getByText('현재 진행 중인 챌린지')).toBeTruthy();
    
    // Check challenge titles are rendered
    expect(getByText('7일간의 감사 일기')).toBeTruthy();
    expect(getByText('30일 긍정 에너지 나누기')).toBeTruthy();
    expect(getByText('21일 명상 습관 만들기')).toBeTruthy();
    
    // Check challenge descriptions are rendered
    expect(getByText('매일 감사한 일 3가지를 기록해보세요.')).toBeTruthy();
    expect(getByText('하루에 한 번 주변 사람에게 긍정적인 말을 해보세요.')).toBeTruthy();
    expect(getByText('매일 10분씩 명상을 하고 느낀 점을 공유해보세요.')).toBeTruthy();
    
    // Check progress percentages are rendered
    expect(getByText('40% 완료')).toBeTruthy();
    expect(getByText('20% 완료')).toBeTruthy();
    expect(getByText('60% 완료')).toBeTruthy();
    
    // Check participant counts are rendered
    const participantTexts = getAllByText(/참여자: \d+명/);
    expect(participantTexts).toHaveLength(3);
    
    // Check durations are rendered
    const durationTexts = getAllByText(/기간: \d+일/);
    expect(durationTexts).toHaveLength(3);
    
    // Check join buttons are rendered
    const joinButtons = getAllByText('참여하기');
    expect(joinButtons).toHaveLength(3);
  });

  it('calls handleJoinChallenge when join button is pressed', () => {
    const { getAllByText } = render(<ChallengeScreen />);
    
    const joinButtons = getAllByText('참여하기');
    
    // Press first join button
    fireEvent.press(joinButtons[0]);
    expect(mockConsoleLog).toHaveBeenCalledWith('Joining challenge:', 1);
    
    // Press second join button
    fireEvent.press(joinButtons[1]);
    expect(mockConsoleLog).toHaveBeenCalledWith('Joining challenge:', 2);
    
    // Press third join button
    fireEvent.press(joinButtons[2]);
    expect(mockConsoleLog).toHaveBeenCalledWith('Joining challenge:', 3);
    
    // Check total calls
    expect(mockConsoleLog).toHaveBeenCalledTimes(3);
  });

  it('renders the ScrollView component', () => {
    const { getByText } = render(<ChallengeScreen />);
    // ScrollView 컴포넌트가 렌더링되면 타이틀이 보여야 함
    expect(getByText('현재 진행 중인 챌린지')).toBeTruthy();
  });

  it('renders cards with content', () => {
    const { getAllByText } = render(<ChallengeScreen />);
    
    // 카드가 렌더링되었는지 제목을 통해 확인
    const challengeTitles = [
      getAllByText('7일간의 감사 일기')[0],
      getAllByText('30일 긍정 에너지 나누기')[0],
      getAllByText('21일 명상 습관 만들기')[0]
    ];
    
    expect(challengeTitles.length).toBe(3);
    
    // 각 카드에 참여하기 버튼이 있는지 확인
    const joinButtons = getAllByText('참여하기');
    expect(joinButtons.length).toBe(3);
  });

  it('displays progress indicators correctly', () => {
    const { getAllByText } = render(<ChallengeScreen />);
    
    // 진행률 텍스트 확인
    const progressTexts = [
      getAllByText('40% 완료')[0],
      getAllByText('20% 완료')[0],
      getAllByText('60% 완료')[0]
    ];
    
    expect(progressTexts.length).toBe(3);
  });
});