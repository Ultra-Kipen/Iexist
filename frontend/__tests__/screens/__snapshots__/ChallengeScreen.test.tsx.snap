// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChallengeScreen creates a snapshot 1`] = `
<RCTScrollView>
  <View>
    <Text>
      현재 진행 중인 챌린지
    </Text>
    <View
      style={
        {
          "marginBottom": 16,
        }
      }
    >
      <View>
        <Text>
          7일간의 감사 일기
        </Text>
        <Text>
          매일 감사한 일 3가지를 기록해보세요.
        </Text>
        <View>
          <View
            accessibilityRole="progressbar"
            progress={0.4}
          />
          <Text>
            40% 완료
          </Text>
        </View>
        <View>
          <Text>
            참여자: 128명
          </Text>
        </View>
        <View>
          <Text>
            기간: 7일
          </Text>
        </View>
      </View>
      <View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <Text>
            참여하기
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        {
          "marginBottom": 16,
        }
      }
    >
      <View>
        <Text>
          30일 긍정 에너지 나누기
        </Text>
        <Text>
          하루에 한 번 주변 사람에게 긍정적인 말을 해보세요.
        </Text>
        <View>
          <View
            accessibilityRole="progressbar"
            progress={0.2}
          />
          <Text>
            20% 완료
          </Text>
        </View>
        <View>
          <Text>
            참여자: 56명
          </Text>
        </View>
        <View>
          <Text>
            기간: 30일
          </Text>
        </View>
      </View>
      <View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <Text>
            참여하기
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        {
          "marginBottom": 16,
        }
      }
    >
      <View>
        <Text>
          21일 명상 습관 만들기
        </Text>
        <Text>
          매일 10분씩 명상을 하고 느낀 점을 공유해보세요.
        </Text>
        <View>
          <View
            accessibilityRole="progressbar"
            progress={0.6}
          />
          <Text>
            60% 완료
          </Text>
        </View>
        <View>
          <Text>
            참여자: 89명
          </Text>
        </View>
        <View>
          <Text>
            기간: 21일
          </Text>
        </View>
      </View>
      <View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <Text>
            참여하기
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
